### PR TITLE
KAFKA-14519; [2/N] New coordinator metrics

### PR DESCRIPTION
**Pull request by jeffkbkim**
This patch copy over existing metrics and add new consumer group metrics to the new GroupCoordinatorService.

Now that each coordinator is responsible for a topic partition, this patch introduces a `GroupCoordinatorMetrics` that records gauges for global metrics such as the number of generic groups in PreparingRebalance state, etc. For `GroupCoordinatorShard` specific metrics, GroupCoordinatorMetrics will activate new `GroupCoordinatorMetricsShard`s that will be responsible for incrementing/decrementing TimelineLong objects and then aggregate the total amount across all shards.

As the CoordinatorRuntime/CoordinatorShard does not care about group metadata, we have introduced a `CoordinatorMetrics.java/CoordinatorMetricsShard.java` so that in the future transaction coordinator metrics can also be onboarded in a similar fashion.

Main files to look at:
* `GroupCoordinatorMetrics.java`
* `GroupCoordinatorMetricsShard.java`
* `CoordinatorMetrics.java`
* `CoordinatorMetricsShard.java`
* `CoordinatorRuntime.java`

Metrics to add after https://github.com/apache/kafka/pull/14408 is merged:
* offset deletions sensor (OffsetDeletions); Meter(offset-deletion-rate, offset-deletion-count)

Metrics to add after https://issues.apache.org/jira/browse/KAFKA-14987 is merged:
 * offset expired sensor (OffsetExpired); Meter(offset-expiration-rate, offset-expiration-count)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: checkstyle/import-control.xml
Comment (jolshan) [inline checkstyle/import-control.xml:252]: I recall we removed yammer metrics from the last pr, do we also want to do that here?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java:139]: for my understanding, this topic partition is only used for metrics tagging?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java]: I noticed we check the metrics shard to ensure it is a GroupCoordinatorMetrics shard in another file. Should we check the type sooner? Should we also check this is the type GroupCoordinatorMetrics (also should the error not include shard here)?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java:554]: Is there a reason we didn't want to put deactivate the shard in onUnloaded?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java]: nit: are my eyes playing tricks on me or is the _ two __?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/CoordinatorMetricsShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/CoordinatorMetricsShard.java]: would it be more useful to have separate methods for global vs not global?

File: checkstyle/import-control.xml
Comment (jolshan) [reply checkstyle/import-control.xml:252]: Or do we have this just to cover the existing generic metrics?

Comment (jeffkbkim) [reply checkstyle/import-control.xml:252]: yeah, unfortunately we need this for the existing metrics

Comment (jolshan) [reply checkstyle/import-control.xml:252]: Are we planning not to add any of the new sensor metrics though? I assume that's what the description is referring to when it says we need to wait for other things to get the sensors.

> Metrics to add after https://github.com/apache/kafka/pull/14408 is merged:
>offset deletions sensor (OffsetDeletions); Meter(offset-deletion-rate, offset-deletion-count)
Metrics to add after https://issues.apache.org/jira/browse/KAFKA-14987 is merged:
>offset expired sensor (OffsetExpired); Meter(offset-expiration-rate, offset-expiration-count)

**Comment by jolshan** [comment]
I took a first pass at the files today, and will take another look more closely at the metrics themselves tomorrow or so.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java:139]: yes

File: checkstyle/import-control.xml
Comment (jeffkbkim) [reply checkstyle/import-control.xml:252]: still working on it, will let you know once it's ready

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java]: :) it actually is 2, and a single underscore is a reserved keyword

Comment (jolshan) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java]: Oh -- java vs scala things 😂

File: checkstyle/import-control.xml
Comment (jeffkbkim) [reply checkstyle/import-control.xml:252]: it's ready

Comment (jolshan) [reply checkstyle/import-control.xml:243]: ooof. I'm too slow at reviewing :(

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java]: do we not want to update this since we are deleting the group anyway?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java]: really minor -- but since the passed in CoordinatorMetrics determines the type of the shard, does it make more sense to check CoordinatorMetrics is of type GroupCoordinatorMetrics and make the error thrown there.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I know we probably can't change the name of the old metric but is it confusing that some generic metrics have the name "generic" and others don't?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java:279]: hmm -- given the states will be different -- should we block this change on the state changes in https://github.com/apache/kafka/pull/14673?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java:138]: in any of these methods do we want to log if the gaugeCounter is null? I'm not sure if we do -- I think there are tradeoffs

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/CoordinatorMetricsShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/CoordinatorMetricsShard.java]: nit: local

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/CoordinatorMetricsShard.java
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/CoordinatorMetricsShard.java]: nit: local

**Comment by jolshan** [comment]
I didn't know where to comment, but do we update the metrics when we load in a shard (ie, all the group states when we loaded etc)

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java]: discussed with david offline; the existing behavior does not record offset deletions if it goes through deleteGroups. so keeping the existing behavior

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java]: The GroupCoordinatorShard is built in CoordinatorRuntime and we don't know which type of CoordinatorMetrics is passed into it. Once we know we're building a GroupCoordinatorShard, only then we can verify the type

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: i agree it's confusing, but not sure if there are alternatives. Would it be better if we rename the sensor variable names at least?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java:279]: https://github.com/apache/kafka/pull/14673 changes the MemberState whereas this PR we only record metrics for ConsumerGroupStates

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java:138]: would this be to catch bugs where the caller does not pass in a valid metric name?

**Comment by jeffkbkim** [comment]
> I didn't know where to comment, but do we update the metrics when we load in a shard (ie, all the group states when we loaded etc)

https://github.com/apache/kafka/pull/14387/files#diff-00f0f81cf13e66781777d94f7d2e68a581663385c37e98792507f2294c91bb09R1356 is where we load in state from the log for consumer groups and

https://github.com/apache/kafka/pull/14387/files#diff-00f0f81cf13e66781777d94f7d2e68a581663385c37e98792507f2294c91bb09R1568 is where we load in state for generic groups.

replay() is called when we read the records from the log

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java
Comment (jolshan) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java:138]: Yeah. Not sure if it is needed, but I guess I'm just used to seeing logging or something when we do null checks. Maybe it's not worth it (or should be very high level logging)

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java
Comment (jolshan) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java:279]: whoops. ok cool so no blocking needed.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java
Comment (jolshan) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java]: I was referring to line 162. At that point we can check the metrics passed in and throw the error.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (jolshan) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Yeah, I wasn't sure if it would be better to just not say "generic" in front, or if that makes it unclear that its the generic group 😅

**Comment by jolshan** [comment]
Also @jeffkbkim can you check the build? Seems like something is off.

**Comment by jolshan** [comment]
> replay() is called when we read the records from the log

Ah got it, so whether we are adding new or loading from the log we will always call replay 👍

**Comment by jeffkbkim** [comment]
> Also @jeffkbkim can you check the build? Seems like something is off.

thanks, i'll take a look

> Ah got it, so whether we are adding new or loading from the log we will always call replay 👍

yup, that's how the new framework works

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java]: ah yes. thanks

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: i updated the code so that all new generic group sensors starts with Group instead of Generic Group

**Comment by jolshan** [comment]
Something very strange is up with this build. I will try again.

**Comment by jolshan** [comment]
Hmm tests are 😅 
I see a lot of `java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: CoordinatorMetrics must be set.`

Did we miss this somewhere?

**Comment by jeffkbkim** [comment]
@jolshan the other test failures look unrelated

**Review by jolshan — APPROVED** [review]
Going to rerun the build to get a better signal, but otherwise lgtm

**Comment by jolshan** [comment]
Build looks good now. I will merge.

**Pull request merged**
